### DISCUSSION TITLE
Using embarrasingly parallel simulations with adjoint solver

### DISCUSSION THREAD

**Discussion by mijna37** [NanoComp/meep#2613] [Q&A] [answered]
Hi, 

I was wondering about using [embarrassing optimizations](https://meep.readthedocs.io/en/latest/Parallel_Meep/#different-forms-of-parallelization) with the adjoint solver. I am optimizing a 2x2 splitter, I understand it makes sense to parallelize the optimizations such that the forward and the adjoint solves are done in parallel since they are totally independent. Is it possible to be used in such a way in MEEP? I have two objective functions J1 and J2 defined. 

My idea is to use `divide_parallel_processes` with 2 independent solvers (one solver for forward, and two for adjoint, since I have two output ports/monitors):

```
n = mp.divide_parallel_processes(3)  
..
#group the gradients 
f0, dJ_du_single = opt([mapping(v, eta, beta)]) 
dJ_du = mp.merge_subgroup_data(dJ_du_single)
```

Does that make sense?

Thanks

**Comment by smartalecH** [comment, marked as answer]
> such that the forward and the adjoint solves are done in parallel since they are totally independent.

Note that the forward and adjoint solves are _not_ independent. You need to run the forward solve first in order to compute the adjoint sources. 

If you have multiple FOMs, you can compute those gradients in parallel as you are describing, but each group must run its forward and adjoint runs sequentially. (Technically the forward run is the same for each FOM... but it's usually a hassle to separate things in that way)
